File: exercise/CompressAlgoBenchmark/cpp/rle.cpp
/**
 * @file rle.cpp
 * @brief Run-Length Encoding (RLE) сжатие и распаковка
 * 
 * Реализация алгоритма RLE для сжатия/распаковки строк.
 * Формат сжатия: "A4B3C2" для "AAAABBBCC"
 * 
 * Особенности:
 * - Работает только с ASCII-символами
 * - Не эффективен для неповторяющихся данных
 */

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>

using namespace std;

/**
 * @brief Сжимает строку с помощью RLE
 * 
 * @param input Входная строка
 * @return string Сжатая строка в формате "A4B3"
 * 
 * @throws invalid_argument Если входная строка пуста
 */
string rle_compress(const string& input) {
    if (input.empty()) {
        throw invalid_argument("Input string cannot be empty");
    }
    
    ostringstream oss;
    size_t i = 0;
    
    while (i < input.size()) {
        char current = input[i];
        size_t count = 1;
        
        while (i + count < input.size() && input[i + count] == current) {
            count++;
        }
        
        oss << current;
        if (count > 1) {
            oss << count;
        }
        
        i += count;
    }
    return oss.str();
}

/**
 * @brief Распаковывает RLE-строку
 * 
 * @param compressed Сжатая строка
 * @return string Исходная строка
 * 
 * @throws invalid_argument При некорректном формате
 */
string rle_decompress(const string& compressed) {
    if (compressed.empty()) {
        throw invalid_argument("Compressed string cannot be empty");
    }
    
    ostringstream oss;
    size_t i = 0;
    
    while (i < compressed.size()) {
        char c = compressed[i++];
        
        if (isdigit(compressed[i])) {
            size_t count = 0;
            while (i < compressed.size() && isdigit(compressed[i])) {
                count = count * 10 + (compressed[i++] - '0');
            }
            oss << string(count, c);
        } else {
            oss << c;
        }
    }
    return oss.str();
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <string>" << endl;
        return 1;
    }
    
    try {
        string original = argv[1];
        auto compressed = rle_compress(original);
        auto decompressed = rle_decompress(compressed);
        
        cout << "Original: " << original << "\n";
        cout << "Compressed: " << compressed << "\n";
        cout << "Decompressed: " << decompressed << "\n";
        cout << "Ratio: " 
             << (1.0 - static_cast<double>(compressed.size()) / original.size()) * 100 
             << "%\n";
    } catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }
    return 0;
}

File: exercise/CompressAlgoBenchmark/python/rle.py
"""
RLE (Run-Length Encoding) Implementation

Features:
- Compresses repeating characters: "AAAABBB" -> "A4B3"
- Handles non-repeating characters: "ABCD" -> "A1B1C1D1"
- Supports ASCII and Unicode
"""

def rle_compress(data: str) -> str:
    """
    Compresses a string using RLE algorithm
    
    Args:
        data: Input string to compress
        
    Returns:
        Compressed string in RLE format
        
    Raises:
        ValueError: If input is empty
    """
    if not data:
        raise ValueError("Input cannot be empty")
    
    compressed = []
    i = 0
    n = len(data)
    
    while i < n:
        char = data[i]
        count = 1
        while i + count < n and data[i + count] == char:
            count += 1
        
        compressed.append(char)
        if count > 1:
            compressed.append(str(count))
        
        i += count
    
    return "".join(compressed)

def rle_decompress(compressed: str) -> str:
    """
    Decompresses RLE-encoded string
    
    Args:
        compressed: RLE-compressed string
        
    Returns:
        Original decompressed string
        
    Raises:
        ValueError: If compressed data is invalid
    """
    if not compressed:
        raise ValueError("Compressed data cannot be empty")
    
    result = []
    i = 0
    n = len(compressed)
    
    while i < n:
        char = compressed[i]
        i += 1
        
        count_str = ""
        while i < n and compressed[i].isdigit():
            count_str += compressed[i]
            i += 1
        
        count = int(count_str) if count_str else 1
        result.append(char * count)
    
    return "".join(result)

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) != 2:
        print("Usage: python rle.py <string>")
        sys.exit(1)
    
    try:
        original = sys.argv[1]
        compressed = rle_compress(original)
        decompressed = rle_decompress(compressed)
        
        print(f"Original: {original}")
        print(f"Compressed: {compressed}")
        print(f"Decompressed: {decompressed}")
        
        ratio = (1 - len(compressed) / len(original)) * 100
        print(f"Compression Ratio: {ratio:.2f}%")
        
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)
